<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Ben Umans" />

<meta name="date" content="2023-09-05" />

<title>for_yunqi</title>

<script src="site_libs/header-attrs-2.14/header-attrs.js"></script>
<script src="site_libs/jquery-3.6.0/jquery-3.6.0.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<style>h1 {font-size: 34px;}
       h1.title {font-size: 38px;}
       h2 {font-size: 30px;}
       h3 {font-size: 24px;}
       h4 {font-size: 18px;}
       h5 {font-size: 16px;}
       h6 {font-size: 12px;}
       code {color: inherit; background-color: rgba(0, 0, 0, 0.04);}
       pre:not([class]) { background-color: white }</style>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/main/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>



<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>

<style type="text/css">code{white-space: pre;}</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>









<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
details > summary > p:only-child {
  display: inline;
}
pre code {
  padding: 0;
}
</style>


<style type="text/css">
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #adb5bd;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script type="text/javascript">
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.tab('show');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');

  // Navbar adjustments
  var navHeight = $(".navbar").first().height() + 15;
  var style = document.createElement('style');
  var pt = "padding-top: " + navHeight + "px; ";
  var mt = "margin-top: -" + navHeight + "px; ";
  var css = "";
  // offset scroll position for anchor links (for fixed navbar)
  for (var i = 1; i <= 6; i++) {
    css += ".section h" + i + "{ " + pt + mt + "}\n";
  }
  style.innerHTML = "body {" + pt + "padding-bottom: 40px; }\n" + css;
  document.head.appendChild(style);
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-bs-toggle="collapse" data-target="#navbar" data-bs-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">For Sharing</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div id="header">



<h1 class="title toc-ignore">for_yunqi</h1>
<h4 class="author">Ben Umans</h4>
<h4 class="date">2023-09-05</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span>
workflowr <span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2023-09-05
</p>
<p>
<strong>Checks:</strong> <span
class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7
<span class="glyphicon glyphicon-exclamation-sign text-danger"
aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>for_sharing/</code> <span
class="glyphicon glyphicon-question-sign" aria-hidden="true"
title="This is the local directory in which the code in this file was executed.">
</span>
</p>
<p>
This reproducible <a href="https://rmarkdown.rstudio.com">R Markdown</a>
analysis was created with <a
  href="https://github.com/workflowr/workflowr">workflowr</a> (version
1.7.0). The <em>Checks</em> tab describes the reproducibility checks
that were applied when the results were created. The <em>Past
versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date
</a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git
repository, you know the exact version of the code that produced these
results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the
global environment can affect the analysis in your R Markdown file in
unknown ways. For reproduciblity it’s best to always run the code in an
empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20230905code">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Seed:</strong>
<code>set.seed(20230905)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20230905code"
class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20230905)</code> was run prior to running
the code in the R Markdown file. Setting a seed ensures that any results
that rely on randomness, e.g. subsampling or permutations, are
reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Session information:</strong>
recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded"
class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package
versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be
confident that you successfully produced the results during this
run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr
project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrong1a75ac0">
<span class="glyphicon glyphicon-ok text-success"
aria-hidden="true"></span> <strong>Repository version:</strong> 1a75ac0
</a>
</p>
</div>
<div id="strongRepositoryversionstrong1a75ac0"
class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development
and connecting the code version to the results is critical for
reproducibility.
</p>
<p>
The results in this page were generated with repository version 1a75ac0.
See the <em>Past versions</em> tab to see a history of the changes made
to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for
the analysis have been committed to Git prior to generating the results
(you can use <code>wflow_publish</code> or
<code>wflow_git_commit</code>). workflowr only checks the R Markdown
file, but you know if there are other scripts or data files that it
depends on. Below is the status of the Git repository when the results
were generated:
</p>
<pre><code>
working directory clean
</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not
included in this status report because it is ok for generated content to
have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were
made to the R Markdown (<code>analysis/for_yunqi.Rmd</code>) and HTML
(<code>docs/for_yunqi.html</code>) files. If you’ve configured a remote
Git repository (see <code>?wflow_git_remote</code>), click on the
hyperlinks in the table below to view the files as they were in that
past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
1a75ac0
</td>
<td>
Ben Umans
</td>
<td>
2023-09-05
</td>
<td>
first commit for upload
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This is to show in some more detail the topics discussed with Yunqi
last week regarding use of ash and mash for looking at eQTL results from
pseudobulked single-cell data.</p>
<p>Data were collected from brain organoids under three conditions (10%
(control), 1%, and 21% ambient oxygen). The single-cell transcriptomes
were classified into cell types, filtered on a variety of quality
metrics, and pseudobulked. MatrixEQTL was used to map cis eQTLs (50 kb
range), using gene expression PCs as covariates, in each cell type and
oxygen condition. Here, I’ll just look at one cell type as an
example.</p>
<pre class="r"><code>library(tidyverse)</code></pre>
<pre><code>── Attaching packages ─────────────────────────────────────── tidyverse 1.3.1 ──</code></pre>
<pre><code>✔ ggplot2 3.4.1     ✔ purrr   1.0.1
✔ tibble  3.2.0     ✔ dplyr   1.1.0
✔ tidyr   1.3.0     ✔ stringr 1.5.0
✔ readr   2.1.4     ✔ forcats 0.5.1</code></pre>
<pre><code>── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()</code></pre>
<pre class="r"><code>library(qvalue)
library(mashr)</code></pre>
<pre><code>Loading required package: ashr</code></pre>
<pre class="r"><code>library(udr)
library(ashr)</code></pre>
<p>Import the results from MatrixEQTL</p>
<pre class="r"><code>matrixeqtl_output &lt;- read_table(file = &quot;/project2/gilad/umans/oxygen_eqtl/data/MatrixEQTL/output/combined_coarse_quality_filter10_072023/results_combined_control10_CyclingProgenitors_nominal.txt&quot;, col_names = TRUE)</code></pre>
<pre><code>
── Column specification ────────────────────────────────────────────────────────
cols(
  snps = col_character(),
  gene = col_character(),
  statistic = col_double(),
  pvalue = col_double(),
  FDR = col_double(),
  beta = col_double()
)</code></pre>
<pre class="r"><code>head(matrixeqtl_output)</code></pre>
<pre><code># A tibble: 6 × 6
  snps       gene  statistic      pvalue    FDR  beta
  &lt;chr&gt;      &lt;chr&gt;     &lt;dbl&gt;       &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt;
1 rs9427117  ADAR      -9.96 0.000000376 0.0172 -1.06
2 rs6688149  ADAR      -9.96 0.000000376 0.0172 -1.06
3 rs3738001  EIF3I     -6.85 0.0000178   0.344  -1.16
4 rs9330261  ADAR      -6.68 0.0000225   0.344  -1.02
5 rs10908422 ADAR      -6.68 0.0000225   0.344  -1.02
6 rs7556080  ADAR      -6.68 0.0000225   0.344  -1.02</code></pre>
<p>There are</p>
<pre class="r"><code>matrixeqtl_output$q &lt;- qvalue(matrixeqtl_output$pvalue)$qvalue

sum(matrixeqtl_output$q&lt;0.05)</code></pre>
<pre><code>[1] 9</code></pre>
<p>significant associations, corresponding to</p>
<pre class="r"><code>matrixeqtl_output[which(matrixeqtl_output$q&lt;0.05),&quot;gene&quot;] %&gt;% unique() %&gt;% nrow()</code></pre>
<pre><code>[1] 2</code></pre>
<p>genes (“eGenes”; UCP2 and ADAR).</p>
</div>
<div id="ash" class="section level2">
<h2>ASH</h2>
<p>Now I use ash on the output from MatrixEQTL, first recovering the SE
from the output statistics.</p>
<pre class="r"><code>matrixeqtl_output &lt;- matrixeqtl_output %&gt;% mutate(se_beta=beta/statistic) 

ash_output &lt;- ash(matrixeqtl_output$beta, matrixeqtl_output$se_beta, mixcompdist = &quot;normal&quot;)</code></pre>
<p>I how have</p>
<pre class="r"><code>cbind(ash_output$result, snps=matrixeqtl_output$snps, gene=matrixeqtl_output$gene) %&gt;% filter(lfsr&lt;0.05) %&gt;% nrow()</code></pre>
<pre><code>[1] 6229</code></pre>
<p>tests with lfsr&lt;0.05, corresponding to</p>
<pre class="r"><code>cbind(ash_output$result, snps=matrixeqtl_output$snps, gene=matrixeqtl_output$gene) %&gt;% filter(lfsr&lt;0.05) %&gt;% pull(gene) %&gt;% unique() %&gt;% length()</code></pre>
<pre><code>[1] 918</code></pre>
<p>eGenes. If I restrict this to only those genes tested in the
three-condition mash case, I get 830 eGenes.</p>
</div>
<div id="mash" class="section level2">
<h2>MASH</h2>
<p>Mash with only a single condition should be equivalent to ash. I
start by using the same input statistics from MatrixEQTL and passing
them to the fastqtl2mash script to focus just on “strong” snp-gene
pairs. I then use this as input for mash. Since there are not multiple
conditions to compare across, I use only the canonical covariance
matrices U.c.</p>
<pre class="r"><code>mash_one_condition_input &lt;- readRDS(&quot;/project2/gilad/umans/oxygen_eqtl/data/MatrixEQTL/output/combined_coarse_quality_filter10_072023/mash/MatrixEQTLSumStats_Cycling10only.mash.rds&quot;)

# https://stephenslab.github.io/udr/reference/ud_init.html
# Use exchangeable zscore model by setting alpha = 1. 
data.random = mash_set_data(as.matrix(mash_one_condition_input$random.b),
                            as.matrix(mash_one_condition_input$random.s), alpha = 0)
data.strong = mash_set_data(as.matrix(mash_one_condition_input$strong.b),
                            as.matrix(mash_one_condition_input$strong.s), alpha = 0)
# why do we use strong.b and not strong.z if alpha=1?
U.c = cov_canonical(data.random)

m = mash(data.random, Ulist = c(U.c), outputlevel = 1)</code></pre>
<pre><code> - Computing 30772 x 127 likelihood matrix.
 - Likelihood calculations took 0.33 seconds.
 - Fitting model with 127 mixture components.
 - Model fitting took 6.07 seconds.</code></pre>
<pre class="r"><code>m2 = mash(data.strong, g=get_fitted_g(m), fixg=TRUE)</code></pre>
<pre><code> - Computing 6858 x 127 likelihood matrix.
 - Likelihood calculations took 0.07 seconds.
 - Computing posterior matrices.
 - Computation allocated took 0.00 seconds.</code></pre>
<p>This yields</p>
<pre class="r"><code>sum(m2$result$lfsr&lt;0.05)</code></pre>
<pre><code>[1] 677</code></pre>
<p>eGenes (as mash reports only the lead snp-eGene pair).</p>
<p>For subsequent comparison, I estimate the mash model again using
alpha=1 (ie, Z-scores), which yields</p>
<pre class="r"><code>mash_one_condition_input &lt;- readRDS(&quot;/project2/gilad/umans/oxygen_eqtl/data/MatrixEQTL/output/combined_coarse_quality_filter10_072023/mash/MatrixEQTLSumStats_Cycling10only.mash.rds&quot;)


# https://stephenslab.github.io/udr/reference/ud_init.html
# Use exchangeable zscore model by setting alpha = 1. 
data.random = mash_set_data(as.matrix(mash_one_condition_input$random.b),
                            as.matrix(mash_one_condition_input$random.s), alpha = 1)
data.strong = mash_set_data(as.matrix(mash_one_condition_input$strong.b),
                            as.matrix(mash_one_condition_input$strong.s), alpha = 1)
# why do we use strong.b and not strong.z if alpha=1?
U.c = cov_canonical(data.random)

m = mash(data.random, Ulist = c(U.c), outputlevel = 1)</code></pre>
<pre><code> - Computing 30772 x 91 likelihood matrix.
 - Likelihood calculations took 0.23 seconds.
 - Fitting model with 91 mixture components.
 - Model fitting took 3.71 seconds.</code></pre>
<pre class="r"><code>m2 = mash(data.strong, g=get_fitted_g(m), fixg=TRUE)</code></pre>
<pre><code> - Computing 6858 x 91 likelihood matrix.
 - Likelihood calculations took 0.05 seconds.
 - Computing posterior matrices.
 - Computation allocated took 0.01 seconds.</code></pre>
<pre class="r"><code>sum(m2$result$lfsr&lt;0.05)</code></pre>
<pre><code>[1] 416</code></pre>
<p>eGenes.</p>
<p>Now I use mash to combine MatrixEQTL output for this same cell type
(Cycling Progenitors) from three different conditions. The “raw”
MatrixEQTL output finds 3 and 0 (1% oxygen condition and 21% oxygen
condition) eGenes with any association test for which q&lt;0.05.</p>
<p>I used the fastqtl2mash again to prepare MatrixEQTL statistics for
mash. Here, I use the Z-scores method in order to take advantage of UDR,
which will yield slightly different results from using the EE model,
which I used above to compare ash to single-condition mash.</p>
<pre class="r"><code>mash_three_condition_input &lt;- readRDS(file = &quot;/project2/gilad/umans/oxygen_eqtl/data/MatrixEQTL/output/combined_coarse_quality_filter10_072023/mash/MatrixEQTLSumStats_CyclingProgenitorsonly.mash.rds&quot;)

# estimate correlation for measurement overlap
data.temp = mash_set_data(mash_three_condition_input$random.b, mash_three_condition_input$random.s)
Vhat = estimate_null_correlation_simple(data.temp)
rm(data.temp)

# https://stephenslab.github.io/udr/reference/ud_init.html
# Use exchangeable zscore model by setting alpha = 1. 
data.random = mash_set_data(mash_three_condition_input$random.b, mash_three_condition_input$random.s, V=Vhat, alpha = 1)
data.strong = mash_set_data(mash_three_condition_input$strong.b, mash_three_condition_input$strong.s, V=Vhat, alpha = 1)
U.c = cov_canonical(data.random)

V.em = mash_estimate_corr_em(data.random, Ulist = c(U.c), details = TRUE)
data.random &lt;- mash_update_data(mashdata = data.random, V = V.em$V, ref = NULL)
data.strong &lt;- mash_update_data(data.strong, V = V.em$V, ref = NULL)

fit0 = ud_init(data.strong, n_rank1 = 0, n_unconstrained = 20)
fit1 = ud_fit(fit0, control = list(unconstrained.update = &quot;ted&quot;, maxiter  = 1e3), verbose=TRUE)
U.ted &lt;- lapply(fit1$U,function (e) &quot;[[&quot;(e,&quot;mat&quot;))
U.c = cov_canonical(data.random)

m = mash(data.random, Ulist = c(U.ted, U.c), outputlevel = 1)
m2 = mash(data.strong, g=get_fitted_g(m), fixg=TRUE)</code></pre>
<p>Now, for the 10% oxygen condition, the mash output yields</p>
<pre class="r"><code>sum(m2$result$lfsr[,&quot;CyclingProgenitors_control10_formash.out&quot;]&lt;0.05)</code></pre>
<pre><code>[1] 468</code></pre>
<p>eGenes.</p>
<p>This would imply that the gain from having two additional conditions
as input yields about 35 additional eGenes, with the vast majority of
the gain over the MatrixEQTL output coming from m/ash. However, mash and
ash seem to disagree substantially. How should I think about this?</p>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span>
Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 4.2.0 (2022-04-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: CentOS Linux 7 (Core)

Matrix products: default
BLAS/LAPACK: /software/openblas-0.3.13-el7-x86_64/lib/libopenblas_haswellp-r0.3.13.so

locale:
 [1] LC_CTYPE=en_US.UTF-8 LC_NUMERIC=C         LC_TIME=C           
 [4] LC_COLLATE=C         LC_MONETARY=C        LC_MESSAGES=C       
 [7] LC_PAPER=C           LC_NAME=C            LC_ADDRESS=C        
[10] LC_TELEPHONE=C       LC_MEASUREMENT=C     LC_IDENTIFICATION=C 

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
 [1] udr_0.3-153     mashr_0.2.57    ashr_2.2-57     qvalue_2.28.0  
 [5] forcats_0.5.1   stringr_1.5.0   dplyr_1.1.0     purrr_1.0.1    
 [9] readr_2.1.4     tidyr_1.3.0     tibble_3.2.0    ggplot2_3.4.1  
[13] tidyverse_1.3.1 workflowr_1.7.0

loaded via a namespace (and not attached):
 [1] fs_1.5.2         lubridate_1.8.0  httr_1.4.5       rprojroot_2.0.3 
 [5] tools_4.2.0      backports_1.4.1  bslib_0.4.1      utf8_1.2.3      
 [9] R6_2.5.1         irlba_2.3.5.1    DBI_1.1.3        colorspace_2.1-0
[13] rmeta_3.0        withr_2.5.0      tidyselect_1.2.0 processx_3.8.0  
[17] compiler_4.2.0   git2r_0.30.1     cli_3.6.0        rvest_1.0.2     
[21] xml2_1.3.3       sass_0.4.2       scales_1.2.1     SQUAREM_2021.1  
[25] mvtnorm_1.1-3    callr_3.7.3      mixsqp_0.3-48    digest_0.6.31   
[29] rmarkdown_2.14   pkgconfig_2.0.3  htmltools_0.5.3  dbplyr_2.1.1    
[33] fastmap_1.1.0    invgamma_1.1     rlang_1.1.0      readxl_1.4.0    
[37] rstudioapi_0.13  jquerylib_0.1.4  generics_0.1.3   jsonlite_1.8.4  
[41] magrittr_2.0.3   Matrix_1.5-3     Rcpp_1.0.10      munsell_0.5.0   
[45] fansi_1.0.4      abind_1.4-5      lifecycle_1.0.3  stringi_1.7.12  
[49] whisker_0.4.1    yaml_2.3.5       plyr_1.8.8       grid_4.2.0      
[53] promises_1.2.0.1 crayon_1.5.2     lattice_0.20-45  haven_2.5.0     
[57] splines_4.2.0    hms_1.1.2        knitr_1.39       ps_1.7.0        
[61] pillar_1.8.1     softImpute_1.4-1 reshape2_1.4.4   reprex_2.0.1    
[65] glue_1.6.2       evaluate_0.15    getPass_0.2-2    modelr_0.1.8    
[69] vctrs_0.6.0      tzdb_0.3.0       httpuv_1.6.5     cellranger_1.1.0
[73] gtable_0.3.1     assertthat_0.2.1 cachem_1.0.6     xfun_0.30       
[77] broom_0.8.0      later_1.3.0      truncnorm_1.0-8  ellipsis_0.3.2  </code></pre>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
https://docs.mathjax.org/en/latest/web/configuration.html. This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>




</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.odd').parent('tbody').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open');
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // temporarily add toc-ignore selector to headers for the consistency with Pandoc
    $('.unlisted.unnumbered').addClass('toc-ignore')

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_');
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
